x/stake/shares.go

```diff
@@ -0,0 +1,139 @@
+}
+
+// amount of bonded shares
+func (s PoolShares) Bonded() sdk.Rat {
+	if s.Kind == ShareBonded {
+		return s.Amount
+	}
+	return sdk.ZeroRat()
+}
+
+//_________________________________________________________________________________________________________
+
+// equivalent amount of shares if the shares were unbonded
+func (s PoolShares) ToUnbonded(p Pool) PoolShares {
+	var amount sdk.Rat
+	switch s.Kind {
+	case ShareBonded:
+		exRate := p.bondedShareExRate().Quo(p.unbondedShareExRate()) // (tok/bondedshr)/(tok/unbondedshr) = unbondedshr/bondedshr
+		amount = s.Amount.Mul(exRate)                                // bondedshr*unbondedshr/bondedshr = unbondedshr
+	case ShareUnbonding:
+		exRate := p.unbondingShareExRate().Quo(p.unbondedShareExRate()) // (tok/unbondingshr)/(tok/unbondedshr) = unbondedshr/unbondingshr
+		amount = s.Amount.Mul(exRate)                                   // unbondingshr*unbondedshr/unbondingshr = unbondedshr
+	case ShareUnbonded:
+		amount = s.Amount
+	}
+	return NewUnbondedShares(amount)
+}
+
+// equivalent amount of shares if the shares were unbonding
+func (s PoolShares) ToUnbonding(p Pool) PoolShares {
+	var amount sdk.Rat
+	switch s.Kind {
+	case ShareBonded:
+		exRate := p.bondedShareExRate().Quo(p.unbondingShareExRate()) // (tok/bondedshr)/(tok/unbondingshr) = unbondingshr/bondedshr
+		amount = s.Amount.Mul(exRate)                                 // bondedshr*unbondingshr/bondedshr = unbondingshr
+	case ShareUnbonding:
+		amount = s.Amount
+	case ShareUnbonded:
+		exRate := p.unbondedShareExRate().Quo(p.unbondingShareExRate()) // (tok/unbondedshr)/(tok/unbondingshr) = unbondingshr/unbondedshr
+		amount = s.Amount.Mul(exRate)                                   // unbondedshr*unbondingshr/unbondedshr = unbondingshr
+	}
+	return NewUnbondingShares(amount)
+}
+
+// equivalent amount of shares if the shares were bonded
+func (s PoolShares) ToBonded(p Pool) PoolShares {
+	var amount sdk.Rat
+	switch s.Kind {
+	case ShareBonded:
+		amount = s.Amount
+	case ShareUnbonding:
+		exRate := p.unbondingShareExRate().Quo(p.bondedShareExRate()) // (tok/ubshr)/(tok/bshr) = bshr/ubshr
+		amount = s.Amount.Mul(exRate)                                 // ubshr*bshr/ubshr = bshr
+	case ShareUnbonded:
+		exRate := p.unbondedShareExRate().Quo(p.bondedShareExRate()) // (tok/ubshr)/(tok/bshr) = bshr/ubshr
+		amount = s.Amount.Mul(exRate)                                // ubshr*bshr/ubshr = bshr
+	}
+	return NewUnbondedShares(amount)
+}
+
+//_________________________________________________________________________________________________________
+
+// get the equivalent amount of tokens contained by the shares
+func (s PoolShares) Tokens(p Pool) sdk.Rat {
+	switch s.Kind {
+	case ShareBonded:
+		return p.unbondedShareExRate().Mul(s.Amount) // (tokens/shares) * shares
+	case ShareUnbonding:
+		return p.unbondedShareExRate().Mul(s.Amount)
+	case ShareUnbonded:
+		return p.unbondedShareExRate().Mul(s.Amount)
+	}
+	return sdk.ZeroRat()
```

Comment:
Should be unreachable code?

Reply:
Yeah maybe this should be a panic on a default scenario instead

Reply:
makes sense